### PR TITLE
Add: Kolmogorov-Smirnov Test

### PR DESCRIPTION
**Pull request by YasinTatarQC**
Hey!

In order to catch distributional shifts in our data-driven projects, I'd like to introduce a [Kolmogorov-Smirnov test](https://en.wikipedia.org/wiki/Kolmogorov%E2%80%93Smirnov_test) to datajudge.

Since this introduces a new `stats.py` subpackage (and this is my first time contributing to datajudge) I'd love to hear your feedback on the following possible issues I see (hence the _draft_ status):

- [x] Afaik, this is the first test that needs the full data column from the database. Therefore, I have introduced `_column(...)` and `get_data(...)` to `db_access`. Is there a better way to do this? `_column(...)` shares a lot with `_column_aggreate(...)` of course.
- [x] Introducing `scipy` as a new dependency just for this test is a bit much. With colleagues, I have discussed options like 
    - (1) just implementing the KS test myself again 
    - (2) stating `scipy` as an optional dependency and warning the user. Any other ideas?
- [x] More tests are needed, in my opinion. Am working on that!

Thanks!

### CODE REVIEW

**Comment by ivergara** [comment]
> Hey!

Hey!

> In order to catch distributional shifts in our data-driven projects, I'd like to introduce a [Kolmogorov-Smirnov test](https://en.wikipedia.org/wiki/Kolmogorov%E2%80%93Smirnov_test) to datajudge.

Thank you for your contribution! I do think it's worth having such constraints being implemented.
 
> Since this introduces a new `stats.py` subpackage (and this is my first time contributing to datajudge) I'd love to hear your feedback on the following possible issues I see (hence the _draft_ status):
> 
> * [ ]  Afaik, this is the first test that needs the full data column from the database. Therefore, I have introduced `_column(...)` and `get_data(...)` to `db_access`. Is there a better way to do this? `_column(...)` shares a lot with `_column_aggreate(...)` of course.

I don't see an issue with having `_column` and `_column_aggregate`, they could be merged and have a keyword flag to control aggregation. But, I don't see it as something really critical. However, I don't see much use for the `get_data` function.

> * [ ]  Introducing `scipy` as a new dependency just for this test is a bit much. With colleagues, I have discussed options like
>   
>   * (1) just implementing the KS test myself again
>   * (2) stating `scipy` as an optional dependency and warning the user. Any other ideas?

Yes, it's a bit much. However, we already have as a "dependency" `snowflake-sqlalchemy` which is rather big too and we can follow the same approach here.

On the other hand, as you state, this constraint implies doing a full copy of the data of two columns locally. I find that somewhat of an issue if we want to handle tables with hundred of millions of rows. 

Now, if you tell me to implement KS yourself, I'd say it'd be a waste to do it in Python so better see if there is a way to implement it at least partially within the confines of the database engine. The challenge here would be to do it in a way that it's database agnostic unless you want to do versions for Postgres, mssql, and snowflake.

Actually, a smarter data access pattern would be highly beneficial even if we keep scipy as the calculator. 

> * [ ]  More tests are needed, in my opinion. Am working on that!

More tests is always good!

**Comment by YasinTatarQC** [comment]
> However, I don't see much use for the `get_data` function.

`get_data` is just there to preserve the current code structure, since you wouldn't want to call _column(...) yourself as a user.
I have now refactored `_column` and removed `_column_aggregate`. Hope that simplifies it a little bit.

**Comment by YasinTatarQC** [comment]
> On the other hand, as you state, this constraint implies doing a full copy of the data of two columns locally. I find that somewhat of an issue if we want to handle tables with hundreds of millions of rows.

Very good point and something I have considered! While Oracle has [KS support built-in ](https://docs.oracle.com/database/121/SQLRF/functions187.htm#SQLRF06319) I couldn't find solutions for MS SQL, Postgres, or Snowflake. There is, however, [this script](https://github.com/spgarbet/sql-kstest) available that would need to be tested.
My solution would be: Check where this script works and create a `ks_supported_databases` list. If our engine is part of this list: nice. If not, warn the user about our scipy dependency and run the test locally on scipy.

File: src/datajudge/db_access.py
Comment (ivergara) [inline src/datajudge/db_access.py:652]: Since you're doing an `else` statement I'd say 

```suggestion
        result =  [elem[0] for elem in result]
```

Together with the next suggestion (couldn't have done it in one block since there is a deleted line in between).

File: src/datajudge/db_access.py
Comment (ivergara) [inline src/datajudge/db_access.py:658]:
```suggestion
    return result, [selection]
```

File: src/datajudge/db_access.py
Comment (ivergara) [inline src/datajudge/db_access.py:635]: I'm not sure the name of this function reflects very well what's doing.

Also, you could properly annotate all the arguments.

Be mindful on how the docstring style is in the rest of the documentation, I don't remember having seen the style with the `@`'s you're using within the codebase.

**Review by ivergara — COMMENTED** [review]
Looks good, but this wasn't the biggest issue in your code. Now in this PR besides implementing your new Constraint, you're adding changes in other places that aren't related. That could have been a separate PR refactoring the existing `_column_aggregate` and the new `_column` together.

**Comment by ivergara** [comment]
> > On the other hand, as you state, this constraint implies doing a full copy of the data of two columns locally. I find that somewhat of an issue if we want to handle tables with hundreds of millions of rows.
> 
> Very good point and something I have considered! While Oracle has [KS support built-in ](https://docs.oracle.com/database/121/SQLRF/functions187.htm#SQLRF06319) I couldn't find solutions for MS SQL, Postgres, or Snowflake. There is, however, [this script](https://github.com/spgarbet/sql-kstest) available that would need to be tested. My solution would be: Check where this script works and create a `ks_supported_databases` list. If our engine is part of this list: nice. If not, warn the user about our scipy dependency and run the test locally on scipy.

Using something like that test would be great. From a fast scan, seems it should be supported by any SQL engine.

Concerning the `ks_supported_databases` list/set/container, I wouldn't worry too much. However, it's fine if you want to add that logic and there is already a precedent of some constraints not working for some supported engines (but it's something we should try to minimize and work in eradicating it). We make the statement we support three engines and we test actively on them, so it's somewhat clear that any use of `datajudge` against other databases is at the user's "risk".

File: src/datajudge/db_access.py
Comment (YasinTatarQC) [reply src/datajudge/db_access.py:635]: What do you think about `_get_column`? I initially wanted to follow the `_column_aggregate` scheme, but this doesn't translate very well, you're right.

Regarding the Docstring: Do we have a pre-configured style? This was the default in PyCharm.
This [post](https://stackoverflow.com/questions/3898572/what-are-the-most-common-python-docstring-formats) talks about different styles and the PEP 257 discussion. In my opinion, `reStructuredText` looks like the default in the community for me. What is your opinion?

Comment (ivergara) [reply src/datajudge/db_access.py:635]: > What do you think about `_get_column`? I initially wanted to follow the `_column_aggregate` scheme, but this doesn't translate very well, you're right.

Yes, probably that conveys better the intention of the function.
> 
> Regarding the Docstring: Do we have a pre-configured style? This was the default in PyCharm. This [post](https://stackoverflow.com/questions/3898572/what-are-the-most-common-python-docstring-formats) talks about different styles and the PEP 257 discussion. In my opinion, `reStructuredText` looks like the default in the community for me. What is your opinion?

I believe we tend to use `numpy` version. Also, if you see most of the code, documentation is mostly absent on the constraints themselves and the documentation basically resides in the `add_xxx` methods of the requirements. I'll defer to @kklein for this.

File: src/datajudge/constraints/stats.py
Comment (kklein) [inline src/datajudge/constraints/stats.py:6]:
```suggestion
from .base import Constraint, OptionalSelections
```
Adapting for consistency. The other modules in `/constraints/` also import relative to the current module.

File: src/datajudge/constraints/stats.py
Comment (kklein) [inline src/datajudge/constraints/stats.py:8]:
```suggestion
from ..db_access import DataReference
```
See above.

File: src/datajudge/constraints/stats.py
Comment (kklein) [inline src/datajudge/constraints/stats.py:7]:
```suggestion
```

File: src/datajudge/constraints/stats.py
Comment (kklein) [inline src/datajudge/constraints/stats.py:25]: Nice. :)

File: src/datajudge/constraints/stats.py
Comment (kklein) [inline src/datajudge/constraints/stats.py:52]:
```suggestion
            f"has p-value {p_value} < {self.significance_level}"
```
The assertion text assumes failure.

File: src/datajudge/db_access.py
Comment (kklein) [inline src/datajudge/db_access.py:650]: Maybe it might make sense to use `.scalars()`instead of `.all()` here. :)
https://docs.sqlalchemy.org/en/14/core/connections.html?highlight=fetchall#sqlalchemy.engine.Result.scalars

File: src/datajudge/requirements.py
Comment (kklein) [inline src/datajudge/requirements.py:1266]: I think something got lost here.

File: tests/integration/test_integration.py
Comment (kklein) [inline tests/integration/test_integration.py:1772]: Should the second argument be `int_table2`? If not, we could remove it from the function parameters.

File: src/datajudge/db_access.py
Comment (kklein) [inline src/datajudge/db_access.py:661]: Could this potentially be called `get_column` - or similar - instead?

File: src/datajudge/db_access.py
Comment (kklein) [reply src/datajudge/db_access.py:635]: > Also, if you see most of the code, documentation is mostly absent on the constraints themselves and the documentation basically resides in the add_xxx methods of the requirements

Yes. The heuristic I've been trying to follow most recently was: 
- Enduser information goes to the `Requirement`'s `add_*` method - this often revolves around what the constraint is about and how to use it.
- Developer information goes into the `Constraint` definition - this often revolves around how the constraint _works_.

**Comment by kklein** [comment]
Hi @YasinTatarQC - thanks a lot for this draft! I'd be very excited to have a ks test be part of datajudge.

I agree with a lot of the things which have been said. Here a few points from my side:
- I think it is not strictly necessary to have the test be done in database, even if this restricts what kind of data the constraint can be used on.
- Yet, I think it would be very cool to have the test be done in database and merely return the 'result' of the test.
- If the test is implemented in database, it would be cool to have a generic, sqlalchemy solution. If this is not in reach, merely getting it to work for postgres, mssql and snowflake is perfectly acceptable from my point of view.

On a different note, I'm currently not sure why the documentation build is failing.

File: tests/integration/test_integration.py
Comment (YasinTatarQC) [reply tests/integration/test_integration.py:1772]: I wanted to test the same data twice to get a perfect p_value of 1.0, and therefore I need the same tables.
But you're right; the second argument is not needed!

File: tests/integration/test_integration.py
Comment (kklein) [inline tests/integration/test_integration.py:1787]: Maybe we can add a parameter set here where the result 'swaps', i.e. on the threshold between sufficient and insufficient statistic value.

**Comment by ivergara** [comment]
> Hi @YasinTatarQC - thanks a lot for this draft! I'd be very excited to have a ks test be part of datajudge.
> 
> I agree with a lot of the things which have been said. Here a few points from my side:
> 
> * I think it is not strictly necessary to have the test be done in database, even if this restricts what kind of data the constraint can be used on.

Upon further discussion, we agreed on continuing doing the "stupid" thing and pulling all the data and doing the computation on Python world. Lets not fall into premature optimization yet, there is precedence of doing the stupid thing in other constraints.

> * Yet, I think it would be very cool to have the test be done in database and merely return the 'result' of the test.

If the decision results in some inconvenient runtime at some point, we can try to think on how to do it smarter. One option is to see if we can do an aggregation step in the database, basically pre-generating the CDF, and then compute locally. That'd reduce massively the data transfer. As a last resort, we have moving the full computation to the database using something akin to the SQL script mentioned earlier. 

However, using that would most likely pose some architectural issues since constraints have the implicit contract of having two datasets when the `test` method is invoked. And in this case, we'd be receiving just one. Anyhow, that's something we can think and talk about once it's really necessary.

File: src/datajudge/constraints/stats.py
Comment (ivergara) [inline src/datajudge/constraints/stats.py:19]:
```suggestion
    def _calculate_2sample_ks_test(data, data2) -> float:
```

You might want to try to annotate `data` and `data2`, but don't think it's critical.

File: src/datajudge/constraints/stats.py
Comment (ivergara) [inline src/datajudge/constraints/stats.py:25]: I'd say that this comment is rather superfluous as it is now. Maybe you could add on why taking this approach and some of its potential failings. And even going further, mentioning at least the first level improvement of letting the engine do the CDF.

File: src/datajudge/constraints/stats.py
Comment (ivergara) [inline src/datajudge/constraints/stats.py:38]: Again, rather superfluous as comment. Seems more to clarify yourself than to help a reader of the future.

File: src/datajudge/db_access.py
Comment (ivergara) [inline src/datajudge/db_access.py:635]: I'd maybe suggest renaming this to `get_column` and get rid of the currently existing `get_column`. It's providing an indirection that's not adding anything since the callable has a default value.

File: src/datajudge/db_access.py
Comment (ivergara) [inline src/datajudge/db_access.py:638]:
```suggestion
    *, aggregate_operator: Callable | None = None,
```
Maybe it'd be good to make this argument keyword-only, which doesn't involve other changes since it's being called always explicitly in all the convenience `get_xx` functions.

File: src/datajudge/requirements.py
Comment (ivergara) [inline src/datajudge/requirements.py:1267]:
```suggestion
        (default is 0.05, i.e., 5%).
```

**Review by ivergara — COMMENTED** [review]
Good job!

Take a look at the few extra observations.

File: src/datajudge/constraints/stats.py
Comment (YasinTatarQC) [reply src/datajudge/constraints/stats.py]: I have added numpy.typing.ArrayLike, though I'm not sure if that's the best approach as I haven't seen it being used in many other places. Makes sense, though, since that is exactly what the scipy method expects (when viewing the docs)

File: src/datajudge/db_access.py
Comment (YasinTatarQC) [reply src/datajudge/db_access.py:635]: I think at this point you may be right. In the future, we might need to add back this indirection if we have more complicated data access patterns, but for now, this additional step is unnecessary perhaps.

Comment (ivergara) [reply src/datajudge/db_access.py]: Sure, once we have the need this could be changed.

File: src/datajudge/constraints/stats.py
Comment (ivergara) [inline src/datajudge/constraints/stats.py:4]: Since this is only needed for typing purposes better wrap it such that doesn't collide with environments not having `numpy`

```python
from typing import TYPE_CHECKING:
    if TYPE_CHECKING:
        from numpy import typing
```

File: src/datajudge/constraints/stats.py
Comment (ivergara) [reply src/datajudge/constraints/stats.py]: I agree that it's in principle a good approach. My only concern is what would happen when someone wants to execute any other constraint in `datajudge` without having `numpy` installed. Will the declaration of the types that don't exist blow up at runtime?

My guess is no, given the `if` guard I suggested for the import. But I have no 100% certainty. What I'm saying is that perhaps you can execute some constraints while removing `numpy` from the environment.

File: src/datajudge/constraints/stats.py
Comment (YasinTatarQC) [reply src/datajudge/constraints/stats.py:20]: Following the guideline of being as generic as possible for inputs, I'd use `Collection` here. A quick look at the scipy source confirms that this should be fine.

Comment (ivergara) [reply src/datajudge/constraints/stats.py:20]: I hoped so too, `scipy` is very lenient on what it accepts as inputs.

File: src/datajudge/constraints/stats.py
Comment (YasinTatarQC) [reply src/datajudge/constraints/stats.py]: I agree with you on all points but I don't think it's worth the effort. We are perfectly fine using Python's `Iterable` or `Collection` type hints.

File: src/datajudge/constraints/stats.py
Comment (kklein) [inline src/datajudge/constraints/stats.py:24]: I don't perfectly understand why we need this even though `scipy` is listed as dependency in `pyproject.toml`. However, if you feel confident about this, please go ahead.

File: tests/integration/test_integration.py
Comment (kklein) [reply tests/integration/test_integration.py:1808]: I think this is outdated.

**Review by kklein — APPROVED** [review]
Great work - thanks a lot! I'm sure this will be a very valuable addition to Datajudge.

Before merging this, it would be cool to add a changelog entry akin to the following

```
1.1.0 - 2022-xx-xx
-------------------

**New feature:**

- Implemented a :class:`~datajudge.BetweenRequirement` constraint running a Kolmogorov Smirnov test.

```

File: src/datajudge/constraints/stats.py
Comment (ivergara) [reply src/datajudge/constraints/stats.py:24]: You're right @kklein , however, the reason why `scipy` is in the `pyproject.toml` file is due to read the docs. Don't ask me why it was failing, but after adding `scipy` here it worked again.

If that weren't the case, this piece of code would be useful to manage the message the user encounters when trying to use this constraint while not having `scipy` installed.

**Comment by ivergara** [comment]
> Great work - thanks a lot! I'm sure this will be a very valuable addition to Datajudge.
> 
> Before merging this, it would be cool to add a changelog entry akin to the following
> 
> ```
> 1.1.0 - 2022-xx-xx
> -------------------
> 
> **New feature:**
> 
> - Implemented a :class:`~datajudge.BetweenRequirement` constraint running a Kolmogorov Smirnov test.
> ```

You can be a bit more verbose than what Kevin suggested.

Also modify the version number in [here](https://github.com/Quantco/datajudge/blob/6be218f1ffec9218ed79567a3e0196bda93fc76c/src/datajudge/__init__.py#L22).

Once this is merged, I'll push the new version in PyPI, and given the addition of `scipy` I might have to go and change the recipe in conda-forge.

**Review by ivergara — APPROVED** [review]

File: CHANGELOG.rst
Comment (ivergara) [inline CHANGELOG.rst:17]: I'd omit this one. It's an internal detail of the library and not part of what a user of this library needs in order to make data tests.

File: CHANGELOG.rst
Comment (ivergara) [inline CHANGELOG.rst:15]:
```suggestion
- Implemented a new constraint :class:`~datajudge...` for :class:`~datajudge.BetweenRequirement` that performs a [Kolmogorov-Smirnov test](https://en.wikipedia.org/wiki/Kolmogorov%E2%80%93Smirnov_test) between two data sources.
```

You'd still need to fix the class reference for the constraint.

**Review by ivergara — CHANGES_REQUESTED** [review]

File: CHANGELOG.rst
Comment (YasinTatarQC) [reply CHANGELOG.rst:17]: Oh okay, I thought the changelog is also directed towards possible developers wanting to extend the library -- either officially or with their own constraints.

Comment (YasinTatarQC) [reply CHANGELOG.rst]: I thought the titles `**title**` are splitting sections and that made it clear. I'm new to `.rst` files, is there any convention we're following or I can lookup?
At first glance, I found [this](https://www.ros.org/reps/rep-0132.html), which is, of course, from a different domain but was nevertheless referenced in a few places.

Comment (YasinTatarQC) [reply CHANGELOG.rst]: Regarding conversation resolving: Sorry! I was cleaning up because I was losing track of some of the tasks. Makes sense of course; basic PR etiquette.

Comment (ivergara) [reply CHANGELOG.rst]: No convention. And indeed you were having sections, but I meant not to just say "new features" or "new method" but to split into a user relevant section and a developer relevant section.

**Review by ivergara — APPROVED** [review]

**Comment by YasinTatarQC** [comment]
Thank you both for the great feedback, learned a lot.

... see you on the next PR. 🎉

**Pull request merged**
